services/cargohold/test/integration/API/V4.hs

```diff
@@ -0,0 +1,282 @@
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+-- This file is part of the Wire Server implementation.
+--
+-- Copyright (C) 2020 Wire Swiss GmbH <opensource@wire.com>
+--
+-- This program is free software: you can redistribute it and/or modify it under
+-- the terms of the GNU Affero General Public License as published by the Free
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module API.V4 (tests) where
+
+import Bilge hiding (body)
+import Bilge.Assert
+import qualified Codec.MIME.Parse as MIME
+import qualified Codec.MIME.Type as MIME
+import Control.Lens hiding (sets)
+import Data.ByteString.Builder
+import qualified Data.ByteString.Char8 as C8
+import Data.ByteString.Conversion
+import qualified Data.ByteString.Lazy as Lazy
+import Data.Id
+import Data.Qualified
+import Data.Text.Encoding (decodeLatin1)
+import Data.Time.Clock
+import Data.Time.Format
+import qualified Data.UUID as UUID
+import Data.UUID.V4
+import Imports hiding (head)
+import Network.HTTP.Client (parseUrlThrow)
+import Network.HTTP.Types.Header
+import Network.HTTP.Types.Method
+import Network.HTTP.Types.Status (status200)
+import Network.Wai.Utilities (Error (label))
+import Test.Tasty
+import Test.Tasty.HUnit
+import TestSetup
+import Wire.API.Asset
+
+tests :: IO TestSetup -> TestTree
+tests s =
+  testGroup
+    "API Integration v4"
+    [ testGroup
+        "simple"
+        [ test s "roundtrip" testSimpleRoundtrip,
+          test s "tokens" testSimpleTokens,
+          test s "s3-upstream-closed" testSimpleS3ClosedConnectionReuse,
+          test s "client-compatibility" testUploadCompatibility
+        ]
+    ]
+
+--------------------------------------------------------------------------------
+-- Simple (single-step) uploads
+
+testSimpleRoundtrip :: TestSignature ()
+testSimpleRoundtrip c = do
+  let def = defAssetSettings
+  let rets = [minBound ..]
+  let sets = def : map (\r -> def & setAssetRetention ?~ r) rets
+  mapM_ simpleRoundtrip sets
+  where
+    simpleRoundtrip sets = do
+      uid <- liftIO $ Id <$> nextRandom
+      uid2 <- liftIO $ Id <$> nextRandom
+      -- Initial upload
+      let bdy = (applicationText, "Hello World")
+      r1 <-
+        uploadSimple (c . path "/assets/v3") uid sets bdy
```

Comment:
is v3 intended here, or perhaps v4?

Reply:
The uploading of assets is still `v3` for users.